lib/IDE/Refactoring.cpp

```diff
@@ -5116,53 +5114,61 @@ class AsyncConverter : private SourceEntityWalker {
       return;
     }
 
-    ParameterList *CallbackParams = Callback->getParameters();
+    ArrayRef<const ParamDecl *> CallbackParams =
+        Callback->getParameters()->getArray();
     ArrayRef<ASTNode> CallbackBody = Callback->getBody()->getElements();
-    if (HandlerDesc.params().size() != CallbackParams->size()) {
+    if (HandlerDesc.params().size() != CallbackParams.size()) {
       DiagEngine.diagnose(CE->getStartLoc(), diag::mismatched_callback_args);
       return;
     }
 
     // Note that the `ErrParam` may be a Result (in which case it's also the
     // only element in `SuccessParams`)
-    ArrayRef<const ParamDecl *> SuccessParams = CallbackParams->getArray();
+    ArrayRef<const ParamDecl *> SuccessParams = CallbackParams;
     const ParamDecl *ErrParam = nullptr;
-    if (HandlerDesc.HasError) {
+    if (HandlerDesc.HasError || HandlerDesc.Type == HandlerType::RESULT) {
```

Comment:
Nitpick: I think this would read better as
```cpp
if (HandlerDesc.Type == HandlerType::RESULT) {
  ErrParam = SuccessParams.back();
} else if (HandlerDesc.HasError) {
  assert(HandlerDesc.Type == HandlerType::PARAMS); // I assume the handler must be valid here, to 100% sure though
  ErrParam = SuccessParams.back();
  SuccessParams = SuccessParams.drop_back();
}
```

It avoids the nesting of checking for `HandlerType::RESULT` and then `HandlerType::PARAMS`.

But feel free to leave it as-is if you like the current spelling better.

Reply:
I don't mind either way, so I've changed it to your suggestion 👍